### PR TITLE
fixes #2165 adds trust domain (network id) configuration/spiffe id look up

### LINKED ISSUE #2165: Add network id

### PR DESCRIPTION
**Pull request by andrewpmartinez**

### CODE REVIEW

File: controller/env/appenv.go [hunk 1]
```diff
@@ -577,11 +580,9 @@ func NewAuthQueryZitiMfa() *rest_model.AuthQueryDetail {
 }
 
 func NewAuthQueryExtJwt(url string) *rest_model.AuthQueryDetail {
-	provider := rest_model.MfaProvidersURL
 	return &rest_model.AuthQueryDetail{
-		HTTPURL:  url,
-		TypeID:   "EXT-JWT",
-		Provider: &provider,
+		HTTPURL: url,
+		TypeID:  "EXT-JWT",
```
Comment (dovholuknf) [inline controller/env/appenv.go:585]: "EXT-JWT" really should be a constant imo and it's used in numerous places already

Comment (andrewpmartinez) [reply controller/env/appenv.go:585]: This wasn't changed and isn't part of the changes I made here.

This is part of a patch I made to get you off the ground a bit back. I'm not sure why its marking this is changed.

File: controller/config.go [hunk 2]
```diff
@@ -313,6 +315,45 @@ func LoadConfig(path string) (*Config, error) {
...
+				trustDomain := value.(string)
+
+				if trustDomain != "" {
+					if !strings.HasPrefix("spiffe://", trustDomain) {
+						trustDomain = "spiffe://" + trustDomain
+					}
+
+					spiffeId, err = url.Parse(trustDomain)
+
+					if err != nil {
```
Comment (dovholuknf) [inline controller/config.go:343]: should this also for invalid input here? `spife://` would pass but has no trust domain, right? does this need to be more robust?

Comment (andrewpmartinez) [reply controller/config.go:343]: Added scheme checking and hostname() == "" checking.

File: controller/config.go [hunk 3]
```diff
@@ -542,6 +583,103 @@ func LoadConfig(path string) (*Config, error) {
...
+				return nil, fmt.Errorf("failed to parse TLS cert at index [%d]: %w", i, err)
+			}
+
+			spiffeId, err := GetSpiffeIdFromCert(cert)
+
+			if err != nil {
+				return nil, fmt.Errorf("failed to determine SPIFFE ID from TLS cert at index [%d]: %w", i, err)
+			}
+
+			if spiffeId != nil {
```
Comment (dovholuknf) [inline controller/config.go:669]: feels like the GetSpiffeIdFromCert just shouldn't allow for nil spiffe id and just be part of the error that retruns from that function? it'd save this particular code from being everhwere, no?

File: controller/config.go [hunk 4]
```diff
@@ -542,6 +583,103 @@ func LoadConfig(path string) (*Config, error) {
...
+
+// GetSpiffeIdFromCert will search a x509 certificate for a trust domain encoded as a spiffe:// URI SAN.
+// Each certificate must contain 0 or 1 spiffe:// URI SAN. The first SPIFFE id looking up the chain is returned. If
+// no SPIFFE id is encountered, nil is returned. Errors are returned for parsing and processing errors only.
+func GetSpiffeIdFromCert(cert *x509.Certificate) (*url.URL, error) {
+	var spiffeId *url.URL
+	for _, uriSan := range cert.URIs {
+		if uriSan.Scheme == "spiffe" {
+			if spiffeId != nil {
+				return nil, fmt.Errorf("multiple URI SAN spiffe:// ids encountered, must only have one")
```
Comment (dovholuknf) [inline controller/config.go:686]: if acceptable, logging the values would be helpful to see when debugging this. just might save me/others some time  to see what's the issue or what the values were that are invalid

**Review by dovholuknf — APPROVED** [review]

File: controller/config.go (hunk 4 above)
Comment (andrewpmartinez) [reply controller/config.go:686]: added id output

File: ziti/cmd/create/config_templates/controller.yml [hunk 5]
```diff
@@ -29,6 +29,8 @@ identity:
   {{ if not .Controller.Web.Identity.AltCertsEnabled }}#{{ end }}  - server_cert:  "{{ .Controller.Web.Identity.AltServerCert }}"
   {{ if not .Controller.Web.Identity.AltCertsEnabled }}#{{ end }}    server_key:   "{{ .Controller.Web.Identity.AltServerKey }}"
 
+trustDomain: quickstart
```
Comment (dovholuknf) [inline ziti/cmd/create/config_templates/controller.yml:32]: this one i'd like to change to an environment variable. i'll offer you a patch to apply if that's ok?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,9 +1,37 @@
+# Release 1.1.5
+
+* Trust Domain Configuration
+
+## Trust Domain Configuration
+
+OpenZiti networks from this release forward will now require a `trust domain` to be configured. High Availability (HA)
+controllers already have this requirement. HA Controllers configure their trust domain via SPIFF ids that are embedded
```
Comment (ekoby) [inline CHANGELOG.md:8]: SPIFFE (sp)?

File: ziti/cmd/create/config_templates/controller.yml (hunk 5 above)
Comment (andrewpmartinez) [reply ziti/cmd/create/config_templates/controller.yml:32]: Lets do it in another PR.

File: CHANGELOG.md (hunk 6 above)
Comment (andrewpmartinez) [reply CHANGELOG.md]: Fixed

File: controller/config.go [hunk 7]
```diff
@@ -313,6 +318,107 @@ func LoadConfig(path string) (*Config, error) {
...
+		if spiffeId == nil {
+			panic("unable to determine a trust domain from a SPIFFE id in the root identity for HA configuration, must have a spiffe:// URI SANs in the server certificate or along the signing CAs chain")
+		}
+	} else {
+		// Non-HA/legacy system, prefer SPIFFE id from certs, but fall back to configuration if necessary
+		spiffeId, _ = GetSpiffeIdFromIdentity(controllerConfig.Id.Identity)
+
+		if spiffeId == nil {
+			//for non HA setups allow the trust domain to come from the configuration root value `trustDomain`
+			if value, found := cfgmap["trustDomain"]; found {
```
Comment (dovholuknf) [inline controller/config.go:340]: it really would be swell to get important constants like "trustDomain" documented and into a constants type location...

Comment (andrewpmartinez) [reply controller/config.go:340]: I had a similar thought; I followed the pattern in the file. I am not going to do it in this PR. If someone actually needs those constants they can go through and do it.

File: controller/config.go [hunk 8]
```diff
@@ -313,6 +318,107 @@ func LoadConfig(path string) (*Config, error) {
...
+	spiffeId.Path = ""
+	controllerConfig.SpiffeIdTrustDomain = spiffeId
+
+	if value, found := cfgmap["additionalTrustDomains"]; found {
+		if valArr, ok := value.([]any); ok {
+			var trustDomains []*url.URL
+			for _, trustDomain := range valArr {
+				if strTrustDomain, ok := trustDomain.(string); ok {
+
+					if !strings.HasPrefix("spiffe://", strTrustDomain) {
```
Comment (dovholuknf) [inline controller/config.go:400]: is spiffee:// case sensitive?

Comment (andrewpmartinez) [reply controller/config.go:400]: I believe the scheme and hostname are both case-insensitive, but I will test them.

Comment (andrewpmartinez) [reply controller/config.go:400]: The scheme is lowercase during parsing. The hostname is not.

**Review by dovholuknf — APPROVED** [review]

**Pull request merged**